addons/udes_security/controllers/main.py

```diff
@@ -5,12 +5,135 @@
+        :param: relation_name: str() of rel table for object and groups
+        :param: inverse_name: str() of inverse relation for rel table to object
+
+        :return: list() res.group ids
+        """
+        query = f"SELECT ARRAY_AGG(gid) FROM {relation_table} WHERE {relation_table}.{inverse_name} IN (%s)"
+        request.env.cr.execute(
+            query,
+            (obj_id,),
+        )
+        result = request.env.cr.fetchone()
+        return result and result[0] or []
+
+    def _get_menu_info_from_action_id(self, action_id):
+        """
+        Get the ir.ui.menu parent_path and actions group_ids from a given action_id
+        `IrUiMenu.sudo().search([("action.id", "=", action_id)])`
+        does not work as `action` is a Reference field. The other option would be to
+        fetch all ir.ui.menu records and then `.filtered()`, but this is a little slow,
+        so instead do a direct query to get only the relevant parts.
+
+        All models on the `action` Reference field on `ir.ui.menu` have the same
+        `_sequence` parameter, so there is no risk of colliding ids from separate models.
+
+        They are stored in the db like "ir.actions.act_window,90" or "ir.actions.server,242"
+        due to being a Reference field, so perform a fuzzy search on %,<id>% to get it
+
+        Be extra safe and ensure `action_id` is an int, as if the user enters text in the URL
+        for action= param, it will come through to here as a string
+
+        :param: action_id: int() for an action (ir.actions.server, ir.actions.act_window, etc..)
+
+        :return: parent_path_ids: list() of ids for all ir.ui.menu records in the hierarchy
+                 action_group_ids:   list() of ids for res.groups for the action
+        """
+        if not isinstance(action_id, int):
+            raise ValidationError("Invalid action ID specified.")
+        query = """
+            SELECT
+                menu.parent_path AS parent_path,
+                field.relation_table AS relation_table,
+                field.column1 AS inverse,
+                SPLIT_PART(menu.action, ',', '2') AS action_id
+            FROM
+                ir_ui_menu menu
+            LEFT JOIN ir_model model ON model.model = SPLIT_PART(menu.action, ',', '1')
+            LEFT JOIN ir_model_fields field ON field.model_id = model.id AND field.name = 'groups_id'
+            WHERE action LIKE '%%,%s'
+        """
+        request.env.cr.execute(query, (action_id,))
+        result = request.env.cr.fetchone()
+        if not result:
+            return [], []
+        (parent_path, relation_table, inverse, action_id) = result
+        # Get ids in order from parent --> child of the menus
+        parent_path_ids = [int(str_menu_id) for str_menu_id in parent_path.split("/")[:-1]]
+        action_group_ids = []
+        if relation_table and inverse:
+            action_group_ids = self._get_object_group_ids(action_id, relation_table, inverse)
```

Comment:
To me it feels like the `if relation_table and inverse` check should be in the function `_get_object_group_ids()` otherwise I guess we would need to add this check before every function call?

Reply:
Maybe, however the function won't do anything without them (it's not possible to find the `group_ids` in sql if we don't have this info). If they were kwargs i'd say yeah but because the function is pointless without them, I think they are best left here.

This is a bit of an edge case handler, where we have an action that isn't bound to a menuitem (it can happen on certain actions) and is really only ever going to be hit from this code which is specifically checking if we couldn't find the reltable/inverse based on the action->menu